### PR TITLE
Feature/kyber aggregator

### PR DESCRIPTION
**Pull request by majkic99**

### CODE REVIEW

File: contracts/exchangeV3/offchainWrappersV3/KyberAggregatorWrapper.sol [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    //Not enough funds
+    error InsufficientFunds(uint256 available, uint256 required);
+    //Not enough eth for protocol fee
+    error InsufficientFeeFunds(uint256 available, uint256 required);
+    //Order success but amount 0
+    error ZeroTokensSwapped();
+
+    using SafeERC20 for IERC20;
+
+    /// @notice Takes order from 0x and returns bool indicating if it is successful
```
Comment (yippee-ki-yay) [inline contracts/exchangeV3/offchainWrappersV3/KyberAggregatorWrapper.sol:31]: Leftover comments referencing 0x

File: contracts/exchangeV3/offchainWrappersV3/KyberAggregatorWrapper.sol [hunk 2]
```diff
@@ -0,0 +1,80 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity =0.8.10;
+
+import "../../utils/SafeERC20.sol";
```
Comment (yippee-ki-yay) [inline contracts/exchangeV3/offchainWrappersV3/KyberAggregatorWrapper.sol:5]: Can remove this import as it's already in DFSExchangeHelper

File: contracts/exchangeV3/offchainWrappersV3/KyberAggregatorWrapper.sol [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    function takeOrder(
+        ExchangeData memory _exData,
+        ExchangeActionType _type
+    ) override public payable returns (bool success, uint256) {
+        // check that contract have enough balance for exchange and protocol fee
+        uint256 tokenBalance = _exData.srcAddr.getBalance(address(this));
+        if (tokenBalance < _exData.srcAmount){
+            revert InsufficientFunds(tokenBalance, _exData.srcAmount);
+        }
+        uint256 ethBalance = TokenUtils.ETH_ADDR.getBalance(address(this));
```
Comment (yippee-ki-yay) [inline contracts/exchangeV3/offchainWrappersV3/KyberAggregatorWrapper.sol:43]: There is not protocolFee in kyber, maybe remove this so not to waste gas

File: scripts/utils/.env.example [hunk 4]
```diff
@@ -0,0 +1,9 @@
+FORK_ID=
```
Comment (yippee-ki-yay) [inline scripts/utils/.env.example:1]: Why have another .env.example here?

File: test/exchange/dfs-sell.js [hunk 5]
```diff
@@ -8,6 +8,6 @@ describe('Dfs-Sell', function () {
 
     it('... should check best price from DFSPrices contract', async () => {
         await dfsSellTest();
-        dfsSellSameAssetTest();
+        // sdfsSellSameAssetTest();
```
Comment (yippee-ki-yay) [inline test/exchange/dfs-sell.js:11]: Can we maybe split up the dfs sell tests, so we have kyber/0x/onchain etc... to be ran separately so we don't have commented out logic in the tests?

File: test/exchange/exchange-tests.js [hunk 6]
```diff
@@ -222,28 +228,125 @@ const dfsSellTest = async () => {
...
             await setNewExchangeWrapper(senderAcc, kyberWrapper.address);
             await setNewExchangeWrapper(senderAcc, uniV3Wrapper.address);
             await setNewExchangeWrapper(senderAcc, curveWrapper.address);
 
             await setNewExchangeWrapper(senderAcc, paraswapWrapper.address);
+            */
+            await setNewExchangeWrapper(senderAcc, kyberAggregatorWrapper.address);
+        });
+
+        it('... should try to sell WETH for DAI with offchain calldata (Kyber)', async () => {
```
Comment (yippee-ki-yay) [inline test/exchange/exchange-tests.js:260]: Try out other token pairs as well in tests?

File: contracts/utils/KyberInputScalingHelper.sol [hunk 7]
```diff
@@ -0,0 +1,345 @@
...
+        uint256 oldAmount,
+        uint256 newAmount
+    ) internal pure returns (bytes memory) {
+        SwapExecutorDescription memory executorDesc = abi.decode(data, (SwapExecutorDescription));
+        executorDesc.minTotalAmountOut = (executorDesc.minTotalAmountOut * newAmount) / oldAmount;
+        for (uint256 i = 0; i < executorDesc.swapSequences.length; i++) {
+            Swap memory swap = executorDesc.swapSequences[i][0];
+            bytes4 functionSelector = swap.functionSelector;
+
+            if (functionSelector == IExecutorHelper.executeUniSwap.selector) {
```
Comment (yippee-ki-yay) [inline contracts/utils/KyberInputScalingHelper.sol:141]: Do we have to update this contract when they add a new exchange?

File: contracts/exchangeV3/offchainWrappersV3/KyberAggregatorWrapper.sol [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+import "../DFSExchangeHelper.sol";
+import "../../interfaces/exchange/IOffchainWrapper.sol";
+import "../../utils/KyberInputScalingHelper.sol";
+
+contract KyberAggregatorWrapper is IOffchainWrapper, DFSExchangeHelper, AdminAuth, DSMath, KyberInputScalingHelper{
+
+    using TokenUtils for address;
+
+    /// @notice offchainData.callData should be this struct encoded
+    struct KyberL1Calldata{
```
Comment (yippee-ki-yay) [inline contracts/exchangeV3/offchainWrappersV3/KyberAggregatorWrapper.sol:17]: No need for struct packing, the calldata can be like in other offchain wrappers, just use array slices from to separate the selector and rest of the call.

Change:
`ExchangeData memory _exData ->  ExchangeData calldata _exData`

And just use:
      `  bytes memory scaledCalldata = getScaledInputData(bytes4(_exData.offchainData.callData[:4]), _exData.offchainData.callData[4:], _exData.srcAmount);`

File: test/exchange/exchange-tests.js [hunk 9]
```diff
@@ -222,28 +228,125 @@ const dfsSellTest = async () => {
...
+                priceObject.routerAddress,
+                allowanceTarget,
+                price,
+                protocolFee,
+                kyberSpecialCalldata,
+            );
+
+            await addToZRXAllowlist(senderAcc, priceObject.routerAddress);
+            // test single action so no changing of amount
+            {
```
Comment (yippee-ki-yay) [inline test/exchange/exchange-tests.js:323]: Put in a separate `it(...)` block don't nest here

**Review by yippee-ki-yay — CHANGES_REQUESTED** [review]
The wrapper seems to be working, but there are a few code quality changes that needs to be done and a bit more cleanup on the tests.

File: contracts/utils/KyberInputScalingHelper.sol (hunk 7 above)
Comment (majkic99) [reply contracts/utils/KyberInputScalingHelper.sol:141]: yes, but it wouldn't be breaking if we put in kyber API to only use these exchanges that are available in contract, and only add new after changing the contract

**Review by yippee-ki-yay — APPROVED** [review]

**Pull request merged**